source/base/mpi.cc

```diff
@@ -812,7 +814,962 @@ namespace Utilities
 #endif
     }
 
+#ifdef DEAL_II_WITH_MPI
+
+    class ConsensusAlgorithmProcess
+    {
+    public:
+      virtual void
+      process_request(const int                                   other_rank,
+                      const std::vector<types::global_dof_index> &buffer_recv,
+                      std::vector<unsigned int> &request_buffer)
+      {
+        (void)other_rank;
+        (void)buffer_recv;
+        (void)request_buffer;
+      }
+
+      virtual std::vector<unsigned int>
+      compute_targets() = 0;
+
+      virtual void
+      pack_data(const int                             other_rank,
+                std::vector<types::global_dof_index> &send_buffer)
+      {
+        (void)other_rank;
+        (void)send_buffer;
+      }
+
+      virtual void
+      prepare_recv_buffer(const int                  other_rank,
+                          std::vector<unsigned int> &recv_buffer)
+      {
+        (void)other_rank;
+        (void)recv_buffer;
+      }
+
+      virtual void
+      unpack_data(const int                        other_rank,
+                  const std::vector<unsigned int> &recv_buffer)
+      {
+        (void)other_rank;
+        (void)recv_buffer;
+      }
+    };
+
+    class ConsensusAlgorithm
+    {
+    public:
+      ConsensusAlgorithm(ConsensusAlgorithmProcess &process,
+                         const MPI_Comm &           comm)
+        : process(process)
+        , comm(comm)
+      {
+        my_rank = this_mpi_process(comm);
+        n_procs = n_mpi_processes(comm);
+      }
+
+      virtual void
+      run() = 0;
+
+    protected:
+      ConsensusAlgorithmProcess &process;
+      const MPI_Comm &           comm;
+      unsigned int               my_rank;
+      unsigned int               n_procs;
+    };
+
+    class ConsensusAlgorithm_NBX : public ConsensusAlgorithm
+    {
+    public:
+      ConsensusAlgorithm_NBX(ConsensusAlgorithmProcess &process,
+                             const MPI_Comm &           comm)
+        : ConsensusAlgorithm(process, comm)
+      {}
+
+      void
+      run() override
+      {
+        start_communication();
+
+        while (!check_own_state())
+          process_requests();
+
+        signal_finish();
+
+        while (!check_global_state())
+          process_requests();
+
+        clean_up_and_end_communication();
+      }
+
+    public:
+      static const unsigned int tag_request  = 12;
+      static const unsigned int tag_delivery = 13;
```

Comment:
Move these two lines to the top of the class for consistency with other classes.